src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.NativeAot.cs

```diff
@@ -1222,9 +1221,23 @@ internal static void ReleaseExternalObjectsFromCurrentThread()
             IntPtr contextToken = GetContextToken();
 
             List<object> objects = new List<object>();
+
             foreach (GCHandle weakNativeObjectWrapperHandle in s_referenceTrackerNativeObjectWrapperCache)
             {
-                ReferenceTrackerNativeObjectWrapper? nativeObjectWrapper = Unsafe.As<ReferenceTrackerNativeObjectWrapper?>(weakNativeObjectWrapperHandle.Target);
+                // Here we aren't part of a GC callback, so other threads can still be running
+                // who are adding and removing from the collection. The collection we are enumerating over
+                // supports lock free enumeration, but it is possible we can run into handles that just
+                // got removed and freed. So we guard against that here by using GCHandle.ToIntPtr
+                // rather than Target given the latter would throw an exception. If they have been freed,
+                // we don't need to worry about them as they don't need to be released.
+                // In addition, we are only handling objects from our current thread here.
+                IntPtr ptr = GCHandle.ToIntPtr(weakNativeObjectWrapperHandle);
```

Comment:
What guarantees that the handle is not freed by some other thread between this check and the `GCHandle.InternalGet` on the next line?

Reply:
Note that `InternalGet` called on freed handle will crash too some of the time. I see that that last commit replaced `IsAllocated` and `Target` with the check for null handle and `InternalGet`. It was not an improvement. Both the previous code and the current look unreliable.

Reply:
> // In addition, we are only handling objects from our current thread here.

This is somewhat misleading if I recall. The "Thread" in the case is the COM context, not the actual thread. This is really looking at STA vs MTA. For STA threads, you perhaps have a chance here, but for an MTA, multiple threads will yield the same answer I believe.

Reply:
I added a lock for this specific function when we iterate over the collection to prevent any modifications to it and thereby any handles being freed.  Hopefully that should address the concerns here.